### PR TITLE
Alerting: support alerting on data.Frame (that can be time series)

### PR DESCRIPTION
**Pull request by kylebrandt**
so new backend plugins can be shimmed into existing alerting

requires https://github.com/grafana/grafana-plugin-sdk-go/pull/95

### CODE REVIEW

**Comment by kylebrandt** [comment]
@bergquist can't you point me to how we wrap errors, you showed me in ... June? But I forgot.

**Comment by kylebrandt** [comment]
for https://github.com/grafana/grafana/issues/22511

**Comment by bergquist** [comment]
As mentioned on slack you can wrap an error like this.
```go
fmt.Errorf("decompress %v: %w", name, err)
```

Personally I dont think that feature is super obvious and I really preferred Dave Cheneys `pkg/errors` https://github.com/pkg/errors That's why we created a small `errutil` package. https://github.com/grafana/grafana/blob/master/pkg/util/errutil/errors.go#L13

```go
errutil.Wrap("A new error", err)
```

Then I went on parental leave and I'm not sure what happened regarding error handling in GO since then :) Time to read up!

File: pkg/tsdb/frame_util.go [hunk 1]
```diff
@@ -35,3 +37,67 @@ func convertTSDBTimePoint(point TimePoint) (t *time.Time, f *float64) {
...
+	timeField := frame.Fields[tsSchema.TimeIndex]
+	timeNullFloatSlice := make([]null.Float, timeField.Len())
+	for i := 0; i < timeField.Len(); i++ {
+		tStamp, err := timeField.FloatAt(i)
+		if err != nil {
+			return nil, err
+		}
+		timeNullFloatSlice[i] = null.FloatFrom(tStamp)
+	}
+	for _, fieldIdx := range tsSchema.ValueIndices {
```
Comment (bergquist) [inline pkg/tsdb/frame_util.go]: Totally code style comment. But can we have some space between those brackets?

File: pkg/services/alerting/conditions/query.go [hunk 2]
```diff
@@ -167,13 +167,26 @@ func (c *QueryCondition) executeQuery(context *alerting.EvalContext, timeRange *
 		if v.Error != nil {
 			return nil, fmt.Errorf("tsdb.HandleRequest() response error %v", v)
 		}
-
-		result = append(result, v.Series...)
+		if v.Dataframes != nil && (v.Series == nil || len(v.Series) == 0) {
```
Comment (bergquist) [inline pkg/services/alerting/conditions/query.go]: This line makes me think too much.

File: pkg/tsdb/frame_util.go (hunk 1 above)
Comment (bergquist) [inline pkg/tsdb/frame_util.go]: Maybe a few comments here? Nested loops can be tricky to read.

File: pkg/tsdb/frame_util.go [hunk 3]
```diff
@@ -35,3 +37,67 @@ func convertTSDBTimePoint(point TimePoint) (t *time.Time, f *float64) {
...
+	return seriesSlice, nil
+
+}
+
+// FramesFromBytes returns a data.Frame slice from marshalled arrow dataframes.
+func FramesFromBytes(bFrames [][]byte) ([]*data.Frame, error) {
+	frames := make([]*data.Frame, len(bFrames))
+	for i, bFrame := range bFrames {
+		var err error
+		frames[i], err = data.UnmarshalArrow(bFrame)
```
Comment (bergquist) [inline pkg/tsdb/frame_util.go]: any reason for not writing?
```go
frames[i], err := data.UnmarshalArrow(bFrame)
```

**Review by bergquist — APPROVED** [review]
I guess this is it. Or am I missing something?

Looks good. Some code style comments but nothing major.

File: pkg/tsdb/frame_util.go (hunk 3 above)
Comment (kylebrandt) [reply pkg/tsdb/frame_util.go]: reason is that it isn't valid go code :-)

Comment (bergquist) [reply pkg/tsdb/frame_util.go]: I was suspecting that. I should have tried it myself. :)

**Pull request merged**